engine/src/main/java/org/terasology/logic/characters/KinematicCharacterMover.java

```diff
@@ -623,105 +625,126 @@ private void walk(final CharacterMovementComponent movementComp, final Character
         Vector3f moveDelta = new Vector3f(endVelocity);
         moveDelta.scale(input.getDelta());
         CharacterCollider collider = movementComp.mode.useCollision ? physics.getCharacterCollider(entity) : null;
+
         MoveResult moveResult = move(state.getPosition(), moveDelta,
                 (state.getMode() != MovementMode.CLIMBING && state.isGrounded() && movementComp.mode.canBeGrounded) ? movementComp.stepHeight : 0,
                 movementComp.slopeFactor, collider);
         Vector3f distanceMoved = new Vector3f(moveResult.getFinalPosition());
         distanceMoved.sub(state.getPosition());
-        state.getPosition().set(moveResult.getFinalPosition());
-        if (input.isFirstRun() && distanceMoved.length() > 0) {
-            entity.send(new MovedEvent(distanceMoved, state.getPosition()));
-        }
-
-        // Upon hitting solid ground, reset the number of jumps back to the maximum value.
-        if (state.isGrounded()) {
-            movementComp.numberOfJumpsLeft = movementComp.numberOfJumpsMax;
-        }
-
-        if (moveResult.isBottomHit()) {
-            if (!state.isGrounded() && movementComp.mode.canBeGrounded) {
-                if (input.isFirstRun()) {
-                    Vector3f landVelocity = new Vector3f(state.getVelocity());
-                    landVelocity.y += (distanceMoved.y / moveDelta.y) * (endVelocity.y - state.getVelocity().y);
-                    logger.debug("Landed at " + landVelocity);
-                    entity.send(new VerticalCollisionEvent(state.getPosition(), landVelocity));
+        if (movementComp.mode == MovementMode.CROUCHING) {
+            if (moveResult.isBottomHit()) {
+                if (state.isGrounded()) {
+                    Vector3f pos = moveResult.getFinalPosition();
+                    Vector3f to = new Vector3f(pos.x, pos.y - 0.35f, pos.z);
+                    SweepCallback callback = collider.sweep(pos, to, VERTICAL_PENETRATION_LEEWAY, -1f);
+                    if (callback.hasHit()) {
+                        state.getPosition().set(moveResult.getFinalPosition());
```

Comment:
instead of setting the state here, the `moveResult` (the movement to perform) should be adjusted to prevent falling off the ledge. this probably needs some better structuring of the whole walk method to get a better overview of things.